---
title: "Assignment 5"
author: "Data Science for Biomedical Informatics (BMIN503/EPID600)"
output:
  html_document:
    toc: false 
    depth: 3 
    theme: paper 
    highlight: tango
---

```{r global options, include = FALSE}
library(tidyverse)
library(sf)
library(tidycensus)
library(leaflet)

knitr::opts_chunk$set(warning = FALSE, message = FALSE)
``` 

***

#### Instructions

- Download the Rmd version of this file
- Complete the questions below in RStudio using the Rmd file as a template 
- Replace text in brackets with your answers, making sure you keep executable code in chunks along with its output to answer the questions as appropriate. (There is no need to keep the brackets).
- Save the Rmd document as Assignment5_*YOUR LAST NAME*.Rmd
- Create an HTML version using knitr
- Turn in completed Assignment5_*YOUR LAST NAME*.html file in Canvas under Assignments -> Assignment 5
- Your assignment **must** be in html format or it will not be graded
- Grades will be assigned according to point scheme below and computed as a percentage of total possible points
- Lateness policy: If an urgent and unforseen circumstance comes up and you need an extension to turn in an assignment, please contact Blanca, Erin, and/or Emma as soon as possible. Unless there is an arrangement in place, late assignments will be scored as follows: 
    - 25% of total score if 1 day late
    - 50% of total score if 2 days late
    - 75%  of total score if 3 days late
    - Assignment will not be graded if 4 or more days late
- DUE DATE: 11/7/19


#### Final Project - Overview, Introduction, Methods/Results

1. Recall that you forked the [Final Project Repo](https://github.com/HimesGroup/BMIN503_Final_Project) and have downloaded it as a project to your local computer. Write the overview and introduction for your final project. The overview consists of 2-3 sentences summarizing the project and goals. For the introduction, the first paragraph describes the problem addressed, its significance, and some background to motivate the problem. In the second paragraph, explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff. Start working on the Methods/Results section, which consists of code and its output along with text describing what you are doing (Note: we will not check your code now, but we encourage you to have something in place before Assignment 6 is distributed).

###Overview
Anti-PD1 therapy has revolutionized cancer treatment however, in hepatocellular carcinoma (HCC) only 20% of respond to treatment.  The mechanisms underlying HCC resistance to anti-PD1 are presently unknown.  This work analyzes HCC RNA sequencing data from the Cancer Genome Atlas (TCGA) to assess whether signatures predictive of anti-PD1 therapy resistance in other tumor types are also present in HCC.

###Introduction
Hepatocellular carcinoma (HCC) is a primary malignancy of liver hepatocytes and the third leading cause of cancer mortality world wide.  In recent years, new systemic therapies have been approved for the treatment of advanced HCC, including immune checkpoint blockade (ICB) targeting programmed cell death protein 1 (PD1).  Unfortunately, only a minority of HCC patients (20%) respond to anti-PD1 therapy and there is little understanding of how anti-PD1 resistance occurs.  Our laboratory has previosly shown in melanoma, breast, and lung cancers that a gene expression signature composed of a subset of interferon (IFN) target genes, termed ISG.RS, predicts acquired resistance to ICB.  Re-analyzing the RNA sequencing data from the TCGA HCC project will offer insights as to whether HCC ICB resistance is mediated by this ISG.RS signature.

Despite precision medicine advances in other tumor types, molecular phenotyping has been limited in HCC as this cancer is typically diagnosed based on radiographic criteria and not biopsied.  However, in 2017, TCGA released molecular profiling including RNA sequencing on 196 HCCs.  This profiling has confirmed the involvement of previously known pathways in HCC pathogenesis, including TP53, beta-catenin, and TERT, however predictive molecular phenotypes of treatment response remain elusive.  Re-analysis of these data crosses multiple disciplines, utilizing gene expression signatures initially described in human data sets and then confirmed mechanistically in mouse models, leveraging RNA sequencing analysis pipelines developed by data scientists, and correlating these findings to clinical care of HCC.  A significant challenge in this analysis is the heterogenity of HCC tumors in the data set in terms of etiology of HCC (hepatitis B, hepatitis C, alcoholic cirrhosis, etc) and tumor grade.  Additionally, these specimens were obtained from surgical resections and may not be representative of tumors which would be exposed to anti-PD1 therapy. However, HCC molecular phenotyping is presently accelerating with a Chinese cohort released in October so additional data sets can be added to this analysis to improve power. 


#### Static Maps

2. Create maps of county-level obesity rate estimates for adults living in the contiguous United States using BRFSS data from 2003 and 2013. The estimates provided have already been age-adjusted using Census population estimates to allow for comparison between counties and across time.
    + Read in [BRFSS obesity data](https://raw.githubusercontent.com/HimesGroup/BMIN503/master/DataFiles/county_obesity_prevalence.csv) and [county polygons](https://raw.githubusercontent.com/HimesGroup/BMIN503/master/DataFiles/uscounties_2010.rds), naming them `obesity` and `counties`, respectively. Use the base _plot_ function to check that `counties` includes the polygon elements you expect. Hint: reading in an RDS file from a website requires that you run the file through a _decompressor_ before loading it via `readRDS`. R has a built-in decomopressor function called `gzcon`. *(2 points)*
    
```{r}
library(tidyverse)

obesity <- read.csv(url("https://raw.githubusercontent.com/HimesGroup/BMIN503/master/DataFiles/county_obesity_prevalence.csv"), header = TRUE)
counties <- readRDS(gzcon(url("https://raw.githubusercontent.com/HimesGroup/BMIN503/master/DataFiles/uscounties_2010.rds")))

```
    
    
    + What were the 2004 and 2014 obesity rates for Orange County, California? For Orange County, Texas? Show all variables associated with these counties in the BRFSS and county polygons datasets. Aside from county names, what identifiers do these datasets share? *(2 points)*
    
    
```{r}
obesity %>%
  filter(county == "Orange County", state =="Texas"|state=="California") 

counties %>%
  filter(NAME == "Orange", STATE==06|STATE==48)

```

Both contain FIPS code but in counties FIPS is seperated by state and county wheras obesity using one FIPS code with both state and county. Counties contains the full GEOID

    + Merge the two datasets so that `counties` contains state names and obesity rates for 2004 and 2014. *(3 points)*
    
```{r}
obesity$fips.code <- sprintf('%05d', obesity$fips.code)
obesity$GEO_ID <- paste0('0500000US', obesity$fips.code)
counties$GEO_ID <- as.character(counties$GEO_ID)
counties <- inner_join(counties, obesity, by = 'GEO_ID')
```
    
    + For each year (i.e., 2004 and 2014), create a static choropleth map of United States county-level obesity rates using _ggplot2_. Add a title with `ggtitle`, remove county borders with `lwd=0` in the `geom_sf` call, and incorporate custom theme elements with the user-created `my_theme()` function. Some code to get you started with these maps is offered below. Feel free to change plot aesthetics or choose a different color palette. *(4 points)*
    
    + How did adult obesity rates change between 2004 and 2014? (Qualitative answer is sufficient!) *(2 points)*    
    
```{r}
library(RColorBrewer)
library(leaflet)

# Use a fixed color scale to more easily compare obesity rates between maps 
counties$age.adjusted.percent.2004 <- as.numeric(as.character(counties$age.adjusted.percent.2004))
counties$age.adjusted.percent.2014 <- as.numeric(as.character(counties$age.adjusted.percent.2014))
counties2<-na.omit(counties)
prev_min <- min(counties$age.adjusted.percent.2004)
prev_max <- max(counties2$age.adjusted.percent.2014) # I know obesity is going up so going to use 2014 which should be higher max

my_theme <- function() {
  theme_minimal() +                                  
  theme(axis.line = element_blank(),                 
        axis.text = element_blank(),                 
        axis.title = element_blank(),
        panel.grid = element_line(color = "white"),  
        legend.key.size = unit(0.8, "cm"),          
        legend.text = element_text(size = 16),       
        legend.title = element_text(size = 16),
        plot.title = element_text(size = 22))      
}

myPalette <- colorRampPalette(brewer.pal(9, "YlOrRd"))

# Replace "..." with your code
ggplot() +
  geom_sf(data = counties, aes(fill = age.adjusted.percent.2004), lwd = 0) +
   my_theme() + 
    ggtitle("Obesity Rate 2004") +
  scale_fill_gradientn(name = "Obesity rate (%)", colours = myPalette(100),
                       limit = range(prev_min, prev_max)) 

ggplot() +
  geom_sf(data = counties, aes(fill = age.adjusted.percent.2014), lwd = 0) +
   my_theme() + 
    ggtitle("Obesity Rate 2014") +
  scale_fill_gradientn(name = "Obesity rate (%)", colours = myPalette(100),
                       limit = range(prev_min, prev_max)) 

```
Adult obesity has dramatically increased between 2004 and 2014 throughout the country.


#### Interactive Maps

3. Create an interactive map to visualize the change in adult obesity rates for all counties in the contiguous United States between 2004 and 2014.
    + Create a new variable in `counties` that tracks the _change_ in obesity rate for each county between 2004 and 2014. Be sure to code this variable so that a positive value indicates an increase in the prevalence of obesity. 
```{r}
counties <- counties %>%
  mutate(obesity.change = age.adjusted.percent.2014 - age.adjusted.percent.2004) 
```
    
    
  
  + Create an interactive choropleth map using the _leaflet_ library to visualize changes in county-level obesity rates between 2004 and 2014. Be sure to include a legend and scalebar for your map. The popup message should include the county name, state name, and change in obesity rate for the selected county. For example, the popup message for Philadelphia, Pennsylvania should read:
    > Philadelphia County, Pennsylvania
    > Change in obesity rate (2004-2014): 
    
    Hint: you can use the _leaflet_ code from the practicum as a starting point. *(5 points)*

```{r}
library(leaflet)

# Bins continuous variables into an ordinal color scale appropriate for our data
pal_fun2 <- colorBin(palette = brewer.pal(9, "RdBu")[c(1:5, 7)], 
                    bins = c(-3, -1, 1, 5, 9, 13, 17), reverse = TRUE,
                    NULL)
pu_message2 <- paste0(counties$county, ", ", counties$state, # paste0 to append tract name with other relevant text 
                     "<br>Change in obesity rate (2004-2014): ",       # <br> forces new line
                     counties$obesity.change, "%")

leaflet(counties) %>%
  addPolygons(stroke = FALSE,                        # remove polygon borders
              fillColor = ~pal_fun2(obesity.change),
              fillOpacity = 0.5, smoothFactor = 0.5, # increase opacity and resolution
              popup = pu_message2) %>%
  addProviderTiles(providers$CartoDB.Positron) %>%   # add third party provider tile
  addLegend("bottomright",                           # location of legend
            pal=pal_fun2,                             # palette function
            values=~obesity.change,                 # variable to be passed to palette function
            title = 'Change in obesity rate 2004-2014',                  # legend title
            opacity = 1) %>%                         # legend opacity (1 = completely opaque)
  addScaleBar()

```

4. Create a choropleth map of a county-aggregated variable of your choice from the American Community Survey (ACS) 5-year estimates for 2012-2016. 
  + Write a line of code to show how you could use a _tidycensus_ command to view variables from the ACS 5-year estimates for 2012-2016. Store the results as an object named `vars`. You do not need to show a preview of `vars`. *(1 point)*
```{r}
library(tidycensus)
vars <- load_variables(dataset = "acs5", 
                       year = 2016) 
```
  
  
  + Use `get_acs` from _tidycensus_ to query a variable of your choice from the ACS 5-year estimates for 2012-2016 at the county level. For percentages, be sure to obtain both an estimate and a total. For full points, assign the ACS data you obtained to a data frame named `acs.data.` Make sure each county is represented by no more than a single row of `acs.data` and each column of `acs.data` contains no more than a single variable and has a meaningful name. We are only interested in mapping the estimate, so you can remove information about margin of error. Show the first six rows of `acs.data`. It is up to you which variable you would like to map. Possible options include:
      + Poverty rate (estimate: B17010_002, total: B17010_001)
      + Median household income (B19013_001)
      + Median house value (B25077_001) *(4 points)* 
        
```{r}
acs.data <- get_acs(geography = 'county',         
                   year = 2016,             
                   variables = c("B19013_001")) # median household income
acs.data <- acs.data %>%
  group_by(GEOID) %>% 
  mutate(med_inc = estimate[variable == "B19013_001"])  %>%
  select(GEOID, NAME, med_inc) 

head(acs.data)

```
        
        
  + Merge `counties` with `acs.data` so that `counties` contains the ACS variables you have chosen to map. Show the first six rows of `counties`. *(2 points)*
  
```{r}
acs.data$GEOID <- paste0('0500000US', acs.data$GEOID)
acs.data <- rename (acs.data, GEO_ID = GEOID)
counties <- inner_join(counties, acs.data, by = 'GEO_ID')

```
  
  
  + Make an interactive map with _leaflet_. Be sure to include a popup message, legend, and scalebar. An example palette function has been provided for you, but feel free to choose a different color palette. For the popup message, be sure to label "$" or "%" as necessary. *(4 points)*
  + Describe in 1-2 sentences the geographic distribution of your chosen variable across the United States. Where is this rate/value the highest? *(1 point)*

    
```{r}
pal_fun <- colorNumeric("BuPu", NULL)       
pu_message <- paste0(counties$county,   
                     "<br>Median Income: ",       # <br> forces new line
                     '$',counties$med_inc)
                                      # default basemap

# Adding more customization 
leaflet(counties) %>%
  addPolygons(stroke = FALSE,                        # remove polygon borders
              fillColor = ~pal_fun(med_inc),
              fillOpacity = 0.5, smoothFactor = 0.5, # increase opacity and resolution
              popup = pu_message) %>%
  addProviderTiles(providers$CartoDB.Positron) %>%   # add third party provider tile
  addLegend("bottomright",                           # location of legend
            pal=pal_fun,                             # palette function
            values=~med_inc,                 # variable to be passed to palette function
            title = 'Median Income',                  # legend title
            opacity = 1) %>%                         # legend opacity (1 = completely opaque)
  addScaleBar()


```
    
Areas of highest median income are concentrated around urban areas, particularly in Northeast and California.   